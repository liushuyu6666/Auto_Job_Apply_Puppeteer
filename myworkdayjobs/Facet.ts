export interface IFacetValue {
    descriptor: string;
    id: string;
    count: number;
}

export interface IFacet {
    facetParameter: string;
    descriptor?: string;
    values: IFacetValue[] | IFacet[];
}

export interface IFacetBrief {
    id: string;
    count: number;
    descriptor: string;
    facetParameter: string;
}

export interface IFacetDetail {
    priority: number;
    values: IFacetBrief[];
}

export type FacetSelector = Map<string, IFacetBrief[]>;

export type FacetSelectorDetail = Map<string, IFacetDetail>;

export type FacetParameterDetail = Map<string, FacetSelectorDetail>;

export class Facet {
    readonly facets: IFacet[];

    constructor(facets: IFacet[]) {
        this.facets = facets;
    }

    private isFacetValue(value: IFacetValue | IFacet): value is IFacetValue {
        return (value as IFacetValue).id !== undefined;
    }

    private dfsValues(
        values: IFacetValue[] | IFacet[],
        facetParameter: string,
        facetSelector: Map<string, IFacetBrief[]>,
    ) {
        for (const value of values) {
            if (this.isFacetValue(value)) {
                this.updateFacetSelector(value, facetParameter, facetSelector);
            } else {
                this.dfsValues(
                    value.values,
                    value.facetParameter,
                    facetSelector,
                );
            }
        }
    }

    private updateFacetSelector(
        value: IFacetValue,
        facetParameter: string,
        facetSelector: Map<string, IFacetBrief[]>,
    ) {
        const facetDescriptor = value.descriptor;
        Array.from(facetSelector.keys()).forEach((loc) => {
            if (facetDescriptor.toLowerCase().indexOf(loc.toLowerCase()) >= 0) {
                facetSelector.get(loc).push({
                    ...value,
                    facetParameter,
                });
            }
        });
    }

    getFacetParameterDetail(
        mainGroupText: string,
        facetPriority: Map<string, number>,
    ): FacetParameterDetail {
        const facetSelector = new Map<string, IFacetBrief[]>();
        Array.from(facetPriority.keys()).forEach((facet) =>
            facetSelector.set(facet, []),
        );

        const mainGroup = this.facets.find(
            (facet) => facet.facetParameter === mainGroupText,
        );

        const values = mainGroup.values;
        this.dfsValues(values, mainGroupText, facetSelector);

        const facetParameterDetail: FacetParameterDetail = new Map();
        const facetSelectorDetail: FacetSelectorDetail = new Map();
        for (const [facet, brief] of facetSelector) {
            const priority = facetPriority.get(facet);
            const facetDetail: IFacetDetail = {
                priority,
                values: brief,
            };
            facetSelectorDetail.set(facet, facetDetail);
        }

        facetParameterDetail.set(mainGroupText, facetSelectorDetail);
        return facetParameterDetail;
    }
}
