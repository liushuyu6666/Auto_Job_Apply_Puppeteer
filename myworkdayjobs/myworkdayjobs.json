[
    {
        "companyName": "loblaw",
        "url": "https://myview.wd3.myworkdayjobs.com/wday/cxs/myview/loblaw_careers/jobs",
        "appliedFacets": {
            "locationRegionStateProvince": ["218a720b28a74c67b5c6d42c00bdadfa"],
            "timeType": ["f9ccda084af243cebebf4538f57811ab"]
        }
    }
]