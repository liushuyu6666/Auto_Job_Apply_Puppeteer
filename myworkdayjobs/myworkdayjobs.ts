import { Browser } from 'puppeteer';
import puppeteer from 'puppeteer-extra';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const StealthPlugin = require('puppeteer-extra-plugin-stealth');

export class MyWorkDayJobs {
    browser: Browser;
    jobSearchPage: string;

    constructor(jobSearchPage: string) {
        this.jobSearchPage = jobSearchPage;
    }

    async init() {
        puppeteer.use(StealthPlugin());
        this.browser = await puppeteer.launch({
            headless: false,
            defaultViewport: null,
        });
    }

    async captureJobSearchPageRequests() {
        try {
            const page = await this.browser.newPage();
            page.on('request', (request) => {
                console.log('Request URL:', request.url());
                console.log(
                    'Request Headers:',
                    JSON.stringify(request.headers(), null, 2),
                );

                // If this is a POST request, you might want to log the POST data
                if (request.method() === 'POST') {
                    console.log('POST Data:', request.postData());
                }
            });

            await page.goto(this.jobSearchPage);
        } catch (error) {
            console.error('Error opening the URL: ', error);
        } finally {
            if (this.browser) await this.browser.close();
        }
    }
}
