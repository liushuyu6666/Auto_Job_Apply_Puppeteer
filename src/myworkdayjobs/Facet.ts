import * as lodash from 'lodash';

export interface IFacetValue {
    descriptor: string;
    id: string;
    count: number;
}

export interface IFacet {
    facetParameter: string;
    descriptor?: string;
    values: IFacetValue[] | IFacet[];
}

export interface IPriority {
    priorityOrder: number;
    priorityText: string;
}

export interface IFacetCard {
    company: string;
    mainGroupText: string;
    facetParameter: string;
    priority: IPriority;
    facetValue: IFacetValue;
}

export class Facet {
    readonly facets: IFacet[];
    readonly company: string;

    constructor(facets: IFacet[], company: string) {
        this.facets = facets;
        this.company = company;
    }

    private isFacetValue(value: IFacetValue | IFacet): value is IFacetValue {
        return (value as IFacetValue).id !== undefined;
    }

    private dfsValues(
        mainGroupText: string,
        facetParameter: string,
        priorities: IPriority[],
        facetValues: IFacetValue[] | IFacet[],
        facetCards: IFacetCard[],
    ) {
        for (const facetValue of facetValues) {
            if (this.isFacetValue(facetValue)) {
                this.updateFacetCards(
                    mainGroupText,
                    facetParameter,
                    priorities,
                    facetValue,
                    facetCards,
                );
            } else {
                this.dfsValues(
                    mainGroupText,
                    facetValue.facetParameter,
                    priorities,
                    facetValue.values,
                    facetCards,
                );
            }
        }
    }

    private updateFacetCards(
        mainGroupText: string,
        facetParameter: string,
        priorities: IPriority[],
        facetValue: IFacetValue,
        facetCards: IFacetCard[],
    ) {
        const excludes = priorities
            .filter((priority) => priority.priorityOrder < 0)
            .map((priority) => priority.priorityText);

        const { descriptor } = facetValue;

        // If descriptor contains the word that need to be excluded
        if (
            excludes.some((exclude) =>
                descriptor.toLowerCase().includes(exclude.toLowerCase()),
            )
        ) {
            return;
        }

        priorities.forEach((priority) => {
            const { priorityText } = priority;
            if (descriptor.toLowerCase().includes(priorityText.toLowerCase())) {
                facetCards.push({
                    company: this.company,
                    mainGroupText,
                    facetParameter,
                    priority,
                    facetValue,
                });
            }
        });
    }

    getPrioritizedFacetCards(
        mainGroupText: string,
        priorities: IPriority[],
    ): IFacetCard[] {
        const facetCards: IFacetCard[] = [];

        const mainGroup = this.facets.find(
            (facet) => facet.facetParameter === mainGroupText,
        );

        const facetParameter = mainGroup.facetParameter;
        const facetValues = mainGroup.values;
        this.dfsValues(
            mainGroupText,
            facetParameter,
            priorities,
            facetValues,
            facetCards,
        );

        return lodash.sortBy(facetCards, ['priority.priorityOrder']);
    }

    getTheFirstPriorityFacetCards(facetCards: IFacetCard[]): IFacetCard[] {
        const minPriorityOrder = facetCards.reduce(
            (prevPriorityOrder, curr) => {
                return prevPriorityOrder > curr.priority.priorityOrder
                    ? curr.priority.priorityOrder
                    : prevPriorityOrder;
            },
            Number.MAX_VALUE,
        );
        return facetCards.filter(
            (card) => card.priority.priorityOrder === minPriorityOrder,
        );
    }
}
