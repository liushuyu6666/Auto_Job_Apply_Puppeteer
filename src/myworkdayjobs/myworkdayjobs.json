{
    "companies": [
        {
            "companyName": "loblaw",
            "apiUrl": "https://myview.wd3.myworkdayjobs.com/wday/cxs/myview/loblaw_careers/jobs",
            "originUrl": "https://myview.wd3.myworkdayjobs.com/en-US/loblaw_careers/"
        },
        {
            "companyName": "walmart",
            "apiUrl": "https://walmart.wd5.myworkdayjobs.com/wday/cxs/walmart/WalmartExternal/jobs",
            "originUrl": "https://walmart.wd5.myworkdayjobs.com/en-US/WalmartExternal/"
        },
        {
            "companyName": "manulife",
            "apiUrl": "https://manulife.wd3.myworkdayjobs.com/wday/cxs/manulife/MFCJH_Jobs/jobs",
            "originUrl": "https://manulife.wd3.myworkdayjobs.com/MFCJH_Jobs"
        },
        {
            "companyName": "myworkdayjobs",
            "apiUrl": "https://workday.wd5.myworkdayjobs.com/wday/cxs/workday/Workday/jobs",
            "originUrl": "https://workday.wd5.myworkdayjobs.com/Workday"
        },
        {
            "companyName": "trendmicro",
            "apiUrl": "https://trendmicro.wd3.myworkdayjobs.com/wday/cxs/trendmicro/External/jobs",
            "originUrl": "https://trendmicro.wd3.myworkdayjobs.com/External"
        },
        {
            "companyName": "bestbuy",
            "apiUrl": "https://bestbuycanada.wd3.myworkdayjobs.com/wday/cxs/bestbuycanada/BestBuyCA_Career/jobs",
            "originUrl": "https://bestbuycanada.wd3.myworkdayjobs.com/BestBuyCA_Career"
        },
        {
            "companyName": "cibc",
            "apiUrl": "https://cibc.wd3.myworkdayjobs.com/wday/cxs/cibc/search/jobs",
            "originUrl": "https://cibc.wd3.myworkdayjobs.com/"
        }
    ],
    "mainGroupPriorities": [
        {
            "mainGroupText": "timeType",
            "priorities": [
                {
                    "priorityOrder": 1,
                    "priorityText": "full time"
                }
            ]
        },
        {
            "mainGroupText": "State__Region__Province",
            "priorities": [
                {
                    "priorityOrder": 1,
                    "priorityText": "Ontario"
                }
            ]
        },
        {
            "mainGroupText": "locationMainGroup",
            "priorities": [
                {
                    "priorityOrder": 1,
                    "priorityText": "Ontario"
                },
                {
                    "priorityOrder": 2,
                    "priorityText": "Canada"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "Vancouver"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "Victoria"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "USA"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "United States"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "Ontario, OH"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "Ontario, OR"
                },
                {
                    "priorityOrder": -1,
                    "priorityText": "Ontario, CA"
                }
            ]
        }
    ]
}
