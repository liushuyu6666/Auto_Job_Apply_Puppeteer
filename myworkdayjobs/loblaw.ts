import mongoose, { Model, Schema } from 'mongoose';
import * as fs from 'fs';
import * as path from 'path';
import * as lodash from 'lodash';
import extractDate from '../utils/extractDate';
import { Facet, IFacet, IPriority } from './Facet';

export interface MyWorkDayJobsPosting {
    title: string;
    externalPath: string;
    locationsText: string;
    postedOn: string;
    bulletFields: string[];
}

// This is the database schema
export interface JobPostings {
    company: string;
    title: string;
    referredId: string;
    portalUrl: string;
    locationsText: string;
    postedOn: Date;
}

export interface Company {
    companyName: string;
    apiUrl: string;
    originUrl: string;
}

export interface MainGroupPriority {
    mainGroupText: string;
    priorities: IPriority[];
}

export interface JobSearchParams {
    companies: Company[];
    mainGroupPriorities: MainGroupPriority[];
}

export type AppliedFacetsValues = Map<string, string[]>;

export interface AppliedFacets {
    appliedFacets: AppliedFacetsValues;
}

export class Loblaw {
    jobPostingModel: Model<JobPostings>;
    filePath: string;

    constructor(filePath: string) {
        const jobPostingSchema = new Schema<JobPostings>({
            company: { type: String, required: true },
            title: { type: String, required: true },
            referredId: { type: String, required: true },
            portalUrl: { type: String, required: true },
            locationsText: { type: String, required: true },
            postedOn: { type: Date, required: true },
        });

        this.jobPostingModel = mongoose.model<JobPostings>(
            'JobPosting',
            jobPostingSchema,
        );

        this.filePath = filePath;
    }

    async fetchInitialFacets(company: string, apiUrl: string): Promise<Facet> {
        const payload = {
            appliedFacets: {},
            limit: 1,
            offset: 0,
        };
        const requestOptions: RequestInit = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        };
        const response = await (await fetch(apiUrl, requestOptions)).json();

        const facets: IFacet[] = response['facets'];
        return new Facet(facets, company);
    }

    async getAppliedFacetsAuto(
        companyName: string,
        apiUrl: string,
        mainGroupPriorities: MainGroupPriority[],
    ): Promise<AppliedFacets> {
        const facet = await this.fetchInitialFacets(companyName, apiUrl);
        const appliedFacetsValues: AppliedFacetsValues = new Map<
            string,
            string[]
        >();

        for (const MainGroupPriority of mainGroupPriorities) {
            const { mainGroupText, priorities } = MainGroupPriority;
            const facetCards = facet.getPrioritizedFacetCards(
                mainGroupText,
                priorities,
            );

            const firstFacetCards =
                facet.getTheFirstPriorityFacetCards(facetCards);
            const grouped = lodash.groupBy(firstFacetCards, 'facetParameter');

            for (const [facetParameter, cards] of Object.entries(grouped)) {
                const ids = cards.map((card) => card.facetValue.id);
                appliedFacetsValues.set(facetParameter, ids);
            }
        }

        return {
            appliedFacets: appliedFacetsValues,
        };
    }

    private async readJobSearchParams(): Promise<JobSearchParams> {
        return JSON.parse(fs.readFileSync(this.filePath, 'utf-8'));
    }

    private async extractAllJobPostings(
        apiUrl: string,
        appliedFacets: AppliedFacets,
    ): Promise<MyWorkDayJobsPosting[]> {
        const allJobPostings: MyWorkDayJobsPosting[] = [];
        let currJobPostings: MyWorkDayJobsPosting[] = [];
        const limit = 20;
        let i = 0;

        // TODO: TOO SLOW
        do {
            const payload = {
                ...appliedFacets,
                limit,
                offset: i * limit,
                searchText: 'software engineer',
            };
            const requestOptions: RequestInit = {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(payload),
            };
            const response = await (await fetch(apiUrl, requestOptions)).json();
            currJobPostings = response['jobPostings'];
            Array.prototype.push.apply(allJobPostings, currJobPostings);
            i++;
        } while (currJobPostings.length === limit && i < 5); // only list top 100 posting

        return allJobPostings;
    }

    private transferToJobPostings(
        jobPostings: MyWorkDayJobsPosting[],
        company: string,
        originUrl: string,
    ): JobPostings[] {
        // 1, Filter
        const jobKeywords = ['software', 'full stack']; // TODO: should be configurable
        const jobKeywordsExclude = ['co-0p']; // TODO: should be configurable
        const filteredJobPostings = jobPostings.filter(
            (jobPosting) =>
                jobKeywords.some(
                    (keyword) =>
                        jobPosting.title.toLowerCase().indexOf(keyword) >= 0,
                ) &&
                !jobKeywordsExclude.some(
                    (keyword) =>
                        jobPosting.title.toLowerCase().indexOf(keyword) >= 0,
                ),
        );

        // 2, Transfer
        const transferredJobPostings: JobPostings[] = [];
        for (const jobPosting of filteredJobPostings) {
            const postedDate = extractDate(jobPosting.postedOn);
            if (postedDate) {
                transferredJobPostings.push({
                    company,
                    title: jobPosting.title,
                    referredId: jobPosting.bulletFields[0],
                    portalUrl: path.join(originUrl, jobPosting.externalPath),
                    locationsText: jobPosting.locationsText,
                    postedOn: postedDate,
                });
            }
        }
        return transferredJobPostings;
    }

    async loadJobPostings(jobPostings: JobPostings[]): Promise<void> {
        for (const jobPosting of jobPostings) {
            const newJobPosting = new this.jobPostingModel(jobPosting);
            await newJobPosting.save();
        }
    }

    /**
     * The entrance function.
     */
    async ETLJobPostings() {
        const { companies, mainGroupPriorities } =
            await this.readJobSearchParams();
        for (const company of companies) {
            const { companyName, originUrl, apiUrl } = company;
            const appliedFacets = await this.getAppliedFacetsAuto(
                companyName,
                apiUrl,
                mainGroupPriorities,
            );
            const allJobPostings = await this.extractAllJobPostings(
                apiUrl,
                appliedFacets,
            );
            const transferredJobPostings = this.transferToJobPostings(
                allJobPostings,
                companyName,
                originUrl,
            );

            await this.loadJobPostings(transferredJobPostings);

            console.log(`finish ${companyName}'s ETL`);
        }
    }
}
