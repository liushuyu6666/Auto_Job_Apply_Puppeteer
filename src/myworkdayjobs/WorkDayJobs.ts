import { Model } from 'mongoose';
import * as fs from 'fs';
import * as path from 'path';
import * as lodash from 'lodash';
import extractDate from '../utils/extractDate';
import { Facet, IFacet, IPriority } from './Facet';
import { JobPosting } from '../models/JobPosting';
import { IJobPostingCount, JobPostingCountModel } from '../models/JobPostingCount';

export interface MyWorkDayJobsPosting {
    title: string;
    externalPath: string;
    locationsText: string;
    postedOn: string;
    bulletFields: string[];
}

export interface Company {
    companyName: string;
    apiUrl: string;
    originUrl: string;
}

export interface MainGroupPriority {
    mainGroupText: string;
    priorities: IPriority[];
}

export interface JobSearchParams {
    companies: Company[];
    mainGroupPriorities: MainGroupPriority[];
}

export type AppliedFacetsValues = Map<string, string[]>;

export interface AppliedFacets {
    appliedFacets: AppliedFacetsValues;
}

export class MyWorkDayJobs {
    filePath: string;
    JobPosting: Model<JobPosting>;
    JobPostingCount: JobPostingCountModel;
    byPostedOn: Map<string, Date>;
    byReferredIds: Map<string, string[]>;

    constructor(filePath: string, JobPostingModel: Model<JobPosting>, JobPostingCountModel: JobPostingCountModel) {
        this.filePath = filePath;
        this.JobPosting = JobPostingModel;
        this.JobPostingCount = JobPostingCountModel;
        this.byPostedOn = new Map<string, Date>();
        this.byReferredIds = new Map<string, string[]>();
    }

    private async init() {
        const latestReferredIds = await this.JobPostingCount.getLatestReferredIds();
        latestReferredIds.forEach((latestReferredId) => {
            this.byPostedOn.set(latestReferredId.company, latestReferredId.postedOn);
            this.byReferredIds.set(latestReferredId.company, latestReferredId.referredIds);
        })
    }

    private async fetchInitialFacets(company: string, apiUrl: string): Promise<Facet> {
        const payload = {
            appliedFacets: {},
            limit: 1,
            offset: 0,
        };
        const requestOptions: RequestInit = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        };
        const response = await (await fetch(apiUrl, requestOptions)).json();

        const facets: IFacet[] = response['facets'];
        return new Facet(facets, company);
    }

    private async getAppliedFacetsAuto(
        companyName: string,
        apiUrl: string,
        mainGroupPriorities: MainGroupPriority[],
    ): Promise<AppliedFacets> {
        const facet = await this.fetchInitialFacets(companyName, apiUrl);
        const appliedFacetsValues: AppliedFacetsValues = new Map<
            string,
            string[]
        >();

        for (const MainGroupPriority of mainGroupPriorities) {
            const { mainGroupText, priorities } = MainGroupPriority;
            const facetCards = facet.getPrioritizedFacetCards(
                mainGroupText,
                priorities,
            );

            const firstFacetCards =
                facet.getTheFirstPriorityFacetCards(facetCards);
            const grouped = lodash.groupBy(firstFacetCards, 'facetParameter');

            for (const [facetParameter, cards] of Object.entries(grouped)) {
                const ids = cards.map((card) => card.facetValue.id);
                appliedFacetsValues.set(facetParameter, ids);
            }
        }

        return {
            appliedFacets: appliedFacetsValues,
        };
    }

    private async readJobSearchParams(): Promise<JobSearchParams> {
        return JSON.parse(fs.readFileSync(this.filePath, 'utf-8'));
    }

    private async extractAllJobPostings(
        apiUrl: string,
        appliedFacets: AppliedFacets,
    ): Promise<MyWorkDayJobsPosting[]> {
        const allJobPostings: MyWorkDayJobsPosting[] = [];
        let currJobPostings: MyWorkDayJobsPosting[] = [];
        const limit = 20;
        let i = 0;

        // TODO: TOO SLOW
        do {
            const payload = {
                appliedFacets: Object.fromEntries(appliedFacets.appliedFacets),
                limit,
                offset: i * limit,
                searchText: 'software engineer',
            };
            const requestOptions: RequestInit = {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(payload),
            };
            const response = await (await fetch(apiUrl, requestOptions)).json();
            currJobPostings = response['jobPostings'];
            Array.prototype.push.apply(allJobPostings, currJobPostings);
            i++;
        } while (currJobPostings.length === limit && i < 5); // only list top 100 posting

        return allJobPostings;
    }

    private async transferToJobPostings(
        jobPostings: MyWorkDayJobsPosting[],
        company: string,
        originUrl: string,
    ): Promise<JobPosting[]> {
        // 1, Filter by titles
        const jobKeywords = ['software', 'full stack']; // TODO: should be configurable
        const jobKeywordsExclude = ['co-op']; // TODO: should be configurable
        const filteredJobPostings = jobPostings.filter(
            (jobPosting) =>
                jobKeywords.some(
                    (keyword) =>
                        jobPosting.title.toLowerCase().indexOf(keyword) >= 0,
                ) &&
                !jobKeywordsExclude.some(
                    (keyword) =>
                        jobPosting.title.toLowerCase().indexOf(keyword) >= 0,
                ),
        );

        // 2, Transfer
        const transferredJobPostings: JobPosting[] = [];
        for (const jobPosting of filteredJobPostings) {
            const postedDate = extractDate(jobPosting.postedOn);
            if (postedDate) {
                transferredJobPostings.push({
                    company,
                    title: jobPosting.title,
                    referredId: jobPosting.bulletFields[0],
                    portalUrl: path.join(originUrl, jobPosting.externalPath),
                    locationsText: jobPosting.locationsText,
                    postedOn: postedDate,
                });
            }
        }

        // 3, Filter by posted date
        const purifiedJobPostings = transferredJobPostings.filter((jobPosting) => {
            const latestPostedOn = this.byPostedOn.get(company);
            const latestReferredIds = this.byReferredIds.get(company);
            if (latestPostedOn && jobPosting.postedOn.getTime() < latestPostedOn.getTime()) {
                return false;
            } else if (latestReferredIds && latestReferredIds.includes(jobPosting.referredId)) {
                return false;
            }
            return true;
        })
    
        return purifiedJobPostings;
    }

    private async loadJobPostings(jobPostings: JobPosting[]): Promise<void> {
        for (const jobPosting of jobPostings) {
            const newJobPosting = new this.JobPosting(jobPosting);
            await newJobPosting.save();
        }
    }

    /**
     * The entrance function.
     */
    async ETLJobPostings() {
        await this.init();

        const { companies, mainGroupPriorities } =
            await this.readJobSearchParams();
        for (const company of companies) {
            const { companyName, originUrl, apiUrl } = company;
            const appliedFacets = await this.getAppliedFacetsAuto(
                companyName,
                apiUrl,
                mainGroupPriorities,
            );

            const allJobPostings = await this.extractAllJobPostings(
                apiUrl,
                appliedFacets,
            );
            const transferredJobPostings = await this.transferToJobPostings(
                allJobPostings,
                companyName,
                originUrl,
            );

            await this.loadJobPostings(transferredJobPostings);

            console.log(`finish ${companyName}'s ETL`);
        }
    }
}
