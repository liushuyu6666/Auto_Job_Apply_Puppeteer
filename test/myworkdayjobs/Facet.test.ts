import { Facet, IFacet, IPriority } from '../../myworkdayjobs/Facet';

const mockFacets: IFacet[] = [
    {
        facetParameter: 'timeType',
        descriptor: 'Time Type',
        values: [
            {
                descriptor: 'Full time',
                id: 'b181d8271e36017533d4ca68eee44f00',
                count: 16207,
            },
            {
                descriptor: 'Part time',
                id: 'b181d8271e3601fa68bcca68eee44e00',
                count: 2136,
            },
        ],
    },
    {
        facetParameter: 'locationMainGroup',
        values: [
            {
                facetParameter: 'locationCountry',
                values: [
                    {
                        facetParameter: 'locationAmericaCountry',
                        values: [
                            {
                                facetParameter: 'locationCountryCanada',
                                values: [
                                    {
                                        descriptor: 'Canada - Hybrid',
                                        id: 'a30a87ed25634629aa6c3958aa2b91ea',
                                        count: 4108,
                                    },
                                    {
                                        descriptor: 'India',
                                        id: 'c4f78be1a8f14da0ab49ce1162348a5e',
                                        count: 249,
                                    },
                                    {
                                        descriptor: 'United States of America',
                                        id: 'bc33aa3152ec42d4995f4791a106ed09',
                                        count: 14820,
                                    },
                                ],
                            },
                        ],
                    },
                    {
                        facetParameter: 'locationIndiaCountry',
                        values: [
                            {
                                facetParameter: 'locationCountryCanada2',
                                values: [
                                    {
                                        descriptor: 'Canada - Remote',
                                        id: 'a30a87ed25634629aa6c3958aa2b91eb',
                                        count: 4108,
                                    },
                                    {
                                        descriptor: 'India',
                                        id: 'c4f78be1a8f14da0ab49ce1162348a5e',
                                        count: 249,
                                    },
                                    {
                                        descriptor: 'United States of America',
                                        id: 'bc33aa3152ec42d4995f4791a106ed09',
                                        count: 14820,
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
            {
                facetParameter: 'locationRegionStateProvince',
                descriptor: 'Location Region/State/Province',
                values: [
                    {
                        descriptor: 'Ontario',
                        id: '745d76f06d2b41738aecee630c5888a0',
                        count: 49,
                    },
                    {
                        descriptor: "(USA) OH ONTARIO 06407 SAM'S CLUB",
                        id: 'e83ebdbd2a0a01fe14d3da8245e9d708',
                        count: 10,
                    },
                ],
            },
            {
                facetParameter: 'primaryLocation',
                descriptor: 'Primary Location',
                values: [
                    {
                        descriptor: 'Texarkana, TX',
                        id: '2ff9989b58280101b1221f3344870000',
                        count: 8,
                    },
                    {
                        descriptor: 'Toronto (Stockyards), ON',
                        id: 'aca8753d865c0101b1ab62fcadce0000',
                        count: 3,
                    },
                ],
            },
        ],
    },
];

describe('Facet', () => {
    let facet: Facet;

    beforeEach(() => {
        facet = new Facet(mockFacets, 'walmart');
    });

    test('getPrioritizedFacetCards retrieves the correct FacetCards, and getTheFirstPriorityFacetCards extract the first priority cards', () => {
        const priorities: IPriority[] = [
            {
                priorityOrder: 1,
                priorityText: 'Toronto',
            },
            {
                priorityOrder: 2,
                priorityText: 'Ontario',
            },
            {
                priorityOrder: 3,
                priorityText: 'Canada',
            },
        ];
        const facetCards = facet.getPrioritizedFacetCards(
            'locationMainGroup',
            priorities,
        );
        const firstCards = facet.getTheFirstPriorityFacetCards(facetCards);

        expect(facetCards).toMatchSnapshot();
        expect(firstCards).toMatchSnapshot();
    });

    test('getPrioritizedFacetCards retrieves the correct FacetCards and excludes USA cards', () => {
        const priorities: IPriority[] = [
            {
                priorityOrder: 1,
                priorityText: 'Toronto',
            },
            {
                priorityOrder: 2,
                priorityText: 'Ontario',
            },
            {
                priorityOrder: 3,
                priorityText: 'Canada',
            },
            {
                priorityOrder: -1,
                priorityText: 'USA',
            },
        ];
        const facetCards = facet.getPrioritizedFacetCards(
            'locationMainGroup',
            priorities,
        );
        const firstCards = facet.getTheFirstPriorityFacetCards(facetCards);

        expect(facetCards).toMatchSnapshot();
        expect(firstCards).toMatchSnapshot();
    });
});
