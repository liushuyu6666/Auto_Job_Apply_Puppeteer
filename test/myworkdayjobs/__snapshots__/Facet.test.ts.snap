// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facet getPrioritizedFacetCards retrieves the correct FacetCards and excludes USA cards 1`] = `
[
  {
    "company": "walmart",
    "facetParameter": "primaryLocation",
    "facetValue": {
      "count": 3,
      "descriptor": "Toronto (Stockyards), ON",
      "id": "aca8753d865c0101b1ab62fcadce0000",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 1,
      "priorityText": "Toronto",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationRegionStateProvince",
    "facetValue": {
      "count": 49,
      "descriptor": "Ontario",
      "id": "745d76f06d2b41738aecee630c5888a0",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 2,
      "priorityText": "Ontario",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationCountryCanada",
    "facetValue": {
      "count": 4108,
      "descriptor": "Canada - Hybrid",
      "id": "a30a87ed25634629aa6c3958aa2b91ea",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 3,
      "priorityText": "Canada",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationCountryCanada2",
    "facetValue": {
      "count": 4108,
      "descriptor": "Canada - Remote",
      "id": "a30a87ed25634629aa6c3958aa2b91eb",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 3,
      "priorityText": "Canada",
    },
  },
]
`;

exports[`Facet getPrioritizedFacetCards retrieves the correct FacetCards and excludes USA cards 2`] = `
[
  {
    "company": "walmart",
    "facetParameter": "primaryLocation",
    "facetValue": {
      "count": 3,
      "descriptor": "Toronto (Stockyards), ON",
      "id": "aca8753d865c0101b1ab62fcadce0000",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 1,
      "priorityText": "Toronto",
    },
  },
]
`;

exports[`Facet getPrioritizedFacetCards retrieves the correct FacetCards, and getTheFirstPriorityFacetCards extract the first priority cards 1`] = `
[
  {
    "company": "walmart",
    "facetParameter": "primaryLocation",
    "facetValue": {
      "count": 3,
      "descriptor": "Toronto (Stockyards), ON",
      "id": "aca8753d865c0101b1ab62fcadce0000",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 1,
      "priorityText": "Toronto",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationRegionStateProvince",
    "facetValue": {
      "count": 49,
      "descriptor": "Ontario",
      "id": "745d76f06d2b41738aecee630c5888a0",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 2,
      "priorityText": "Ontario",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationRegionStateProvince",
    "facetValue": {
      "count": 10,
      "descriptor": "(USA) OH ONTARIO 06407 SAM'S CLUB",
      "id": "e83ebdbd2a0a01fe14d3da8245e9d708",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 2,
      "priorityText": "Ontario",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationCountryCanada",
    "facetValue": {
      "count": 4108,
      "descriptor": "Canada - Hybrid",
      "id": "a30a87ed25634629aa6c3958aa2b91ea",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 3,
      "priorityText": "Canada",
    },
  },
  {
    "company": "walmart",
    "facetParameter": "locationCountryCanada2",
    "facetValue": {
      "count": 4108,
      "descriptor": "Canada - Remote",
      "id": "a30a87ed25634629aa6c3958aa2b91eb",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 3,
      "priorityText": "Canada",
    },
  },
]
`;

exports[`Facet getPrioritizedFacetCards retrieves the correct FacetCards, and getTheFirstPriorityFacetCards extract the first priority cards 2`] = `
[
  {
    "company": "walmart",
    "facetParameter": "primaryLocation",
    "facetValue": {
      "count": 3,
      "descriptor": "Toronto (Stockyards), ON",
      "id": "aca8753d865c0101b1ab62fcadce0000",
    },
    "mainGroupText": "locationMainGroup",
    "priority": {
      "priorityOrder": 1,
      "priorityText": "Toronto",
    },
  },
]
`;
